src/cunumeric/stat/histogram.cu

```diff
@@ -0,0 +1,235 @@
+std::tuple<size_t, const VAL*> get_accessor_ptr(const AccessorRO<VAL, 1>& src_acc,
+                                                const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// RD accessor (size, pointer) extractor:
+//
+template <typename VAL>
+std::tuple<size_t, VAL*> get_accessor_ptr(const AccessorRD<SumReduction<VAL>, true, 1>& src_acc,
+                                          const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// accessor copy utility:
+//
+template <typename VAL>
+std::tuple<size_t, Buffer<VAL>, const VAL*> make_accessor_copy(const AccessorRO<VAL, 1>& src_acc,
+                                                               const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size      = src_rect.hi - src_rect.lo + 1;
+  Buffer<VAL> src_copy = create_buffer<VAL>(src_size, Legion::Memory::Kind::GPU_FB_MEM);
+  return std::make_tuple(src_size, src_copy, src_ptr);
+}
+}  // namespace detail
+
+template <Type::Code CODE>
+struct HistogramImplBody<VariantKind::GPU, CODE> {
+  using VAL = legate_type_of<CODE>;
+
+  // for now, it has been decided to hardcode these types:
+  //
+  using BinType    = double;
+  using WeightType = double;
+
+  // in the future we might relax relax that requirement,
+  // but complicate dispatching:
+  //
+  // template <typename BinType = VAL, typename WeightType = VAL>
+  void operator()(const AccessorRO<VAL, 1>& src,
+                  const Rect<1>& src_rect,
+                  const AccessorRO<BinType, 1>& bins,
+                  const Rect<1>& bins_rect,
+                  const AccessorRO<WeightType, 1>& weights,
+                  const Rect<1>& weights_rect,
+                  const AccessorRD<SumReduction<WeightType>, true, 1>& result,
+                  const Rect<1>& result_rect) const
+  {
+    auto stream          = get_cached_stream();
+    cudaStream_t stream_ = static_cast<cudaStream_t>(stream);
+
+    auto&& [src_size, src_copy, src_ptr] = detail::make_accessor_copy(src, src_rect);
+    CHECK_CUDA(cudaMemcpyAsync(
+      src_copy.ptr(0), src_ptr, src_size * sizeof(VAL), cudaMemcpyDeviceToDevice, stream));
```

Comment:
I think you can use `thrust::copy` here, then the code will be even closer to the CPU versions, so you could potentially have a single version of this code for all 3 variants!

Reply:
That sounds pretty cool, actually. Let me check if `thrust::copy()` might be sync, in which case we could not use it instead. I will get back to this.

Reply:
Done according to the 4 points we agreed:

1. unify accessor helpers to one undifferentiated form for every ecosystem (GPU/CPU/OMP);
2. put them in a header (*_gen.h), under a deeper nested namespace, say, cunumeric::detail::accessor namespace;
3. move the copy inside make_accessor_copy();
4. do the copying via thrust::copy_n();